### PR TITLE
Breakers usage and other improvements

### PR DESCRIPTION
**Pull request by saneshark**
## Description of change
We had an outage in MVI, I'm adding these specs to get a better handle on how we respond to these types of outages with respect to breakers. In so doing, I identified several other controllers that are doing things incorrectly.

1. Breakers needs to appear at the top of the middleware stack if the Statsd plugin is to capture the request / response times correctly. There are a handful of places where it is not the top most middleware.

2. Breakers only responds to certain error classes. If those specific errors are not raised somewhere in the middleware, then circuit breakers wont work as expected, however you should still use the breakers gem because of how the StatsD plugin keeps track of latency metrics.

3. In several places we are using `conn.use` where the proper thing to do is to use `conn.response` or `conn.request` instead. The Timeout middleware should be more explicit and not designed as a use middleware. Additionally, it should raise an error class that Breakers can properly respond to with respect to an outage.

4. MVI makes 12 external requests on development and 10 external requests on test. I need to find out why that is. I think it's because several places in User.rb, we invoke classes that then turn around and load instances of user.rb again, and since we do not cache MVI for failures and record_not_found, additional external requests get made. This is a huge CODE SMELL, and probably the explanation for why latency on /users tends to be so poor. I would like to introduce memcached to handle that variety of caching, distinct from what we are doing in Redis since it is a cache specific to the current request that the user made. I don't think it is a good idea to mix these two styles of caching together. **UPDATE This will be addressed in separate PR.**

5. Ive suppressed or fixed some STDOUT output that has become bothersome: The json-schema-validation warnings will be fixed by bumping the gem version thanks to the contribution that @jholton made in that repository fixing the issue. Additionally, I've removed the output from the WebBrick thread that remove_cookie middleware spews out. That should help with some of the spurious warnings we get and help to draw attention to warnings that actually matter when running our test suite.

## Testing done
Just specs

## Testing planned
Just more specs

### CODE REVIEW

File: spec/request/user_request_spec.rb [hunk 1]
```diff
@@ -95,4 +95,73 @@
...
+    let(:loa3_user) { build(:user, :loa3) }
+    let(:auth_header) { { 'Authorization' => "Token token=#{token}" } }
+
+    before do
+      Session.create(uuid: loa3_user.uuid, token: token)
+      User.create(loa3_user)
+      create(:account, idme_uuid: loa3_user.uuid)
+    end
+
+    it 'GET /v0/user - for MVI error should only make a request to MVI one time per request!' do
```
Comment (annaswims) [inline spec/request/user_request_spec.rb]: This fails 
````
   it 'GET /v0/user - for MVI error should only make a request to MVI one time per request!' do
      stub_mvi_failure
      expect_any_instance_of(Common::Client::Base).to receive(:perform).twice.and_call_original
      get v0_user_url, nil, auth_header
      get v0_user_url, nil, auth_header
    end
````

Comment (annaswims) [reply spec/request/user_request_spec.rb]: never mind.  @saneshark explained that `expect_any_instance_of` doesn't work like that.

File: lib/mvi/configuration.rb [hunk 2]
```diff
@@ -47,15 +47,15 @@ def ssl_options
 
     def connection
       Faraday.new(base_path, headers: base_request_headers, request: request_options, ssl: ssl_options) do |conn|
+        conn.use :breakers
```
Comment (annaswims) [inline lib/mvi/configuration.rb]: I think a comment explaining how the order matters is worth it.

File: lib/central_mail/configuration.rb [hunk 3]
```diff
@@ -15,7 +15,7 @@ def connection
         faraday.request :multipart
         faraday.request :url_encoded
 
-        faraday.use :breakers
+        faraday.use :breakers # FIXME: breakers must appear first, to work correctly.
```
Comment (saneshark) [inline lib/central_mail/configuration.rb]: @lihanli

File: lib/emis/configuration.rb [hunk 4]
```diff
@@ -25,10 +25,11 @@ def ssl_options
 
     def connection
       Faraday.new(base_path, headers: base_request_headers, request: request_options, ssl: ssl_options) do |conn|
+        # FIXME: don't use a `use` middleware if its only for requests.
         conn.use Common::Client::Middleware::Request::RescueTimeout, backend_service: :emis
```
Comment (saneshark) [inline lib/emis/configuration.rb]: @jholton

File: lib/hca/configuration.rb [hunk 5]
```diff
@@ -52,12 +52,14 @@ def ssl_options
 
     def connection
       Faraday.new(base_path, headers: base_request_headers, request: request_options, ssl: ssl_options) do |conn|
+        # FIXME: don't use a `use` middleware if it is only for requests.
         conn.use Common::Client::Middleware::Request::RescueTimeout, { backend_service: :hca }, 'api.hca.timeout'
```
Comment (saneshark) [inline lib/hca/configuration.rb]: @jholton

File: lib/hca/configuration.rb [hunk 6]
```diff
@@ -52,12 +52,14 @@ def ssl_options
...
         conn.options.open_timeout = 10  # TODO(molson): Make a config/setting
         conn.options.timeout = 15       # TODO(molson): Make a config/setting
         conn.request :soap_headers
         conn.use HCA::SOAPParser
-        conn.use :breakers
+        conn.use :breakers # FIXME: breakers must appear first, to work correctly.
+        # FIXME: Don't use a `use` middleware if it is only for responses.
         conn.use Common::Client::Middleware::Response::RescueTimeout, { backend_service: :hca }, 'api.hca.timeout'
```
Comment (saneshark) [inline lib/hca/configuration.rb]: @jholton

File: lib/hca/configuration.rb [hunk 7]
```diff
@@ -52,12 +52,14 @@ def ssl_options
...
     def connection
       Faraday.new(base_path, headers: base_request_headers, request: request_options, ssl: ssl_options) do |conn|
+        # FIXME: don't use a `use` middleware if it is only for requests.
         conn.use Common::Client::Middleware::Request::RescueTimeout, { backend_service: :hca }, 'api.hca.timeout'
         conn.options.open_timeout = 10  # TODO(molson): Make a config/setting
         conn.options.timeout = 15       # TODO(molson): Make a config/setting
         conn.request :soap_headers
         conn.use HCA::SOAPParser
-        conn.use :breakers
+        conn.use :breakers # FIXME: breakers must appear first, to work correctly.
```
Comment (saneshark) [inline lib/hca/configuration.rb]: @lihanli

File: lib/preneeds/configuration.rb [hunk 8]
```diff
@@ -37,7 +37,7 @@ def connection
         conn.response :eoas_xml_errors
         conn.response :clean_response
 
-        conn.use :breakers
+        conn.use :breakers # FIXME: breakers must appear first, to work correctly.
```
Comment (saneshark) [inline lib/preneeds/configuration.rb]: @lihanli

File: lib/preneeds/logged_configuration.rb [hunk 9]
```diff
@@ -18,8 +18,8 @@ def connection
         conn.response :eoas_xml_errors
         conn.response :clean_response
 
-        conn.use :breakers
-        conn.use :logging, 'PreneedsBurial'
+        conn.use :breakers # FIXME: breakers must appear first, to work correctly.
+        conn.use :logging, 'PreneedsBurial' # FIXME: `use` middleware should appear (above `request` middleware).
```
Comment (saneshark) [inline lib/preneeds/logged_configuration.rb]: @lihanli

**Comment by saneshark** [comment]
@lihanli will you be merging changes for the items I tagged you on in separate PRs or contributing to this one? Either way works for me. I'll just update this to resolve conflicts as problems identified are addressed.

**Comment by lihanli** [comment]
i can update this one, the first change was made before i saw this pr

File: lib/common/client/base.rb [hunk 10]
```diff
@@ -34,6 +37,13 @@ def connection
             raise SecurityError, 'http client needs cookies stripped'
           end
 
+          if handlers.include?(Breakers::UptimeMiddleware)
+            return unless handlers.first == Breakers::UptimeMiddleware
+            raise BreakersImplementationError, 'Breakers should be the first middleware implemented.'
+          else
+            warn('Breakers is not implemented for this service.')
+          end
```
Comment (saneshark) [inline lib/common/client/base.rb]: Thanks for the suggestion @patrickvinograd

**Comment by saneshark** [comment]
@jholton thanks for the contribution! It looks good. Might get it merged faster if you did a separate PR. And I can just upmerge here once you do. Otherwise it sounds good to me. I think if we can get the breakers moved up in the middlewares stack in those couple of places, this will be ready for merge.

File: lib/common/client/base.rb (hunk 10 above)
Comment (annaswims) [reply lib/common/client/base.rb]: This is great!

File: lib/common/client/middleware/logging.rb [hunk 11]
```diff
@@ -19,11 +19,26 @@ def call(env)
...
+        # as an object in your hash when setting response_env middleware lower in the stack.
+        def handle_response_bodies(body)
+          if body.is_a?(Ox::Document)
+            Ox.dump(body)
+          elsif body.is_a?(Hash)
+            body.to_json
+          else
+            body
+          end
+        end
```
[the remark this answers is not in the document]
Comment (jkassemi) [reply lib/common/client/middleware/logging.rb]: this seems fine - only concern for me would be backwards incompatibility with debug scripts, but I don't know of any active ongoing debug efforts.

[the remark this answers is not in the document]
Comment (dhall) [reply lib/common/client/middleware/logging.rb]: I can speak to the mvi usage of this: I do any debugging manually from rails console so this wouldn't effect me negatively, in fact it's a nice improvement :)

**Pull request merged**